src/core/common/modulecontainer.cpp

```diff
@@ -0,0 +1,273 @@
+    return util::contains_if(loaded, [&](const auto& lib) {
+        std::error_code ec;
+        return std::filesystem::equivalent(path, lib, ec);
+    });
+};
+
+void ModuleContainer::unload() {
+    factoryObject_.reset();
+    sharedLibrary_.reset();
+    if (std::filesystem::is_regular_file(tmpFile_)) {
+        std::filesystem::remove(tmpFile_);
+    }
+}
+void ModuleContainer::load(bool runtimeReload) {
+    if (runtimeReload && !isLoaded(libFile_)) {
+        auto tmpFile = getTmpDir() / libFile_.filename();
+        // Load a copy of the file to make sure that we can overwrite the
+        // file.
+        std::error_code ec;
+        if (!std::filesystem::copy_file(libFile_, tmpFile,
+                                        std::filesystem::copy_options::update_existing, ec)) {
+            throw Exception(IVW_CONTEXT, "Unable to write temporary file {} since: {}", tmpFile,
+                            ec.message());
+        }
+        tmpFile_ = tmpFile;
+    }
+
+    filesystem::setWorkingDirectory(filesystem::getInviwoBinDir());
+    sharedLibrary_ = std::make_unique<SharedLibrary>(tmpFile_.empty() ? libFile_ : tmpFile_);
+
+    if (auto moduleFunc = sharedLibrary_->findSymbolTyped<f_getModule>("createModule")) {
+        factoryObject_.reset(moduleFunc());
+    } else {
+        throw Exception(
+            IVW_CONTEXT,
+            "Could not find 'createModule' function needed for creating the module in {}. "
+            "Make sure that you have compiled the library and exported the function.",
+            libFile_);
+    }
+
+    identifier_ = toLower(factoryObject_->name);
+    protectedModule_ = factoryObject_->protectedModule == ProtectedModule::on;
+    protectedLibrary_ = protectedModule_;
+}
+
+ModuleContainer::ModuleContainer(ModuleContainer&&) = default;
+ModuleContainer& ModuleContainer::operator=(ModuleContainer&&) = default;
+
+ModuleContainer::~ModuleContainer() {
+    resetModule();
+    unload();
+}
+const std::string& ModuleContainer::identifier() const { return identifier_; }
+const std::string& ModuleContainer::name() const { return factoryObject_->name; }
+
+void ModuleContainer::setReloadCallback(InviwoApplication* app,
+                                        std::function<void(ModuleContainer&)> callback) {
+    if (!libFile_.empty()) {
+        observer_ = std::make_unique<LibFileObserver>(
+            app, libFile_, [this](const std::filesystem::path&) { reloadCallback_(*this); });
+    }
+}
+
+void ModuleContainer::createModule(InviwoApplication* app) {
+    if (!module_) {
+        // This is a huge hack since we cant pass this information into the InviwoModule constructor
+        // without updating all the module constructors which we want to avoid.
+        // And we need the inforation in the constructor since we use it to register various paths
+        // in the constructor. like python script and glsl code
+        app->getModuleManager().setModuleLocator([this](const InviwoModule& m) {
```

Comment:
```suggestion
        // This is a huge hack since we can't pass this information into the InviwoModule constructor
        // without updating all the module constructors, which we want to avoid.
        // And we need the information in the constructor since we use it to register various paths
        // in the constructor, like python script and glsl code.
        app->getModuleManager().setModuleLocator([this](const InviwoModule& m) {
```